File: 2022/03.js
const R = require('ramda')
const readline = require('readline')
const rl = readline.createInterface({ input: process.stdin })

const lines = []

rl.on('line', data => {
  const line = [
    data.slice(0, data.length / 2), // .split(''),
    data.slice(data.length / 2), // .split(''),
  ]

  lines.push(line)
})

rl.on('close', () => {
  console.log('partOne', partOne(lines))
  console.log('partTwo', partTwo(lines))
})

const prio = s => (s === s.toUpperCase())
  ? s.charCodeAt(0) - 'A'.charCodeAt(0) + 27
  : s.charCodeAt(0) - 'a'.charCodeAt(0) + 1

function partOne (lines) {
  return R.sum(
    R.map(
      R.pipe(
        ([left, right]) => left.split('').find(c => right.includes(c)),
        prio,
      ),
    )(lines))
}

function partTwo (lines) {
  return 'todo'
}

File: 2022/03.py
import sys
import string


def prio(char):
    return 1 + string.ascii_letters.index(char)


def part_one(lines):
    result = 0
    for line in lines:
        half = len(line) // 2
        item, = set(line[:half]) & set(line[half:])
        result += prio(item)

    return result


def part_two(lines):
    result = 0
    for i in range(0, len(lines), 3):
        item, = set(lines[i]) & set(lines[i+1]) & set(lines[i+2])
        result += prio(item)

    return result


def main():
    lines = []
    for line in sys.stdin:
        line = line.replace('\n', '')
        lines.append(line)

    print('part_one', part_one(lines))

    print('part_two', part_two(lines))


main()
